src/include/art.h

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2024 The pgmoneta community
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list
+ * of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this
+ * list of conditions and the following disclaimer in the documentation and/or other
+ * materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without specific
+ * prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PGMONETA_ART_H
+#define PGMONETA_ART_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <deque.h>
+
+#include <stdint.h>
+
+#define MAX_PREFIX_LEN 10
+
+enum art_node_type {
+   Node4,
+   Node16,
+   Node48,
+   Node256
+};
+
+typedef int (*art_callback)(void* data, const unsigned char* key, uint32_t key_len, void* value);
+
+typedef void (*value_destroy_callback)(void* value);
+
+/**
+ * This struct is included as part
+ * of all the various node sizes.
+ * All node types should be aligned
+ * because we need the last bit to be 0 as a flag bit for leaf.
+ * So that leaf can be treated as a node as well and stored in the children field,
+ * and only converted back when necessary
+ */
+struct art_node
+{
+   uint32_t prefix_len;                     /**< The actual length of the prefix segment */
+   enum art_node_type type;                 /**< The node type */
+   uint8_t num_children;                    /**< The number of children */
+   unsigned char prefix[MAX_PREFIX_LEN];    /**< The (potentially partial) prefix, only record up to MAX_PREFIX_LEN characters */
+} __attribute__ ((aligned (64)));
```

Comment:
we don't need alignment

Reply:
We do. That makes sure that a pointer to an art node struct always has 0 in the least significant bit. And if a node is leaf we set the bit to 1 as an indicator. See SET_LEAF in art.c